engineering/architecture/numbers.md

```diff
@@ -0,0 +1,106 @@
+
+- When the front end cannot parse the canonical stringified number, it falls back to a simplistic formatter which only performs string transformation to ensure only that the decimal separator matches the user's locale. This produces a formatted result which lacks grouping separators, partially sacrificing Goal D.
+
+### UX
+
+- The front-end allows users to enter the Postgres numbers `NaN`, `-Infinity`, and `Infinity` using a context menu -- and it displays those numbers stylized (similar to `NULL`).
+
+- While the user is entering a number, Mathesar _partially_ reformats the user's entry in order to match the final display of the number as closely as possible (to satisfy Goal E). For example, with a "en-US" locale, the entry `1234` is reformatted to `1,234` as the user types. However, some entries cannot be _fully_ reformatted as the user types. For example, an entry of `1.` would be fully formatted to `1` but must preserve the trailing decimal so that the user may type additional characters.
+
+    After the user shifts focus away from the number input, Mathesar _fully_ reformats their entry. For example, `1.` will be reformatted to `1` (removing the trailing zero).
+
+### Scientific notation
+
+- To satisfy Goal F, Mathesar delegates to the front-end the responsibility to decide between displaying a number in standard notation or scientific notation.
+
+    The algorithm for making this decision is as follows.
+    
+    - If the front-end cannot parse the number, then it retains the notation used in the stringified number.
+    - If the front-end can only produce a `BigInt`, then it will use standard notation because: (a) `Intl.NumberFormat` is not capable of accurately formatting very large `BigInt` values due to its upper limit on `maximumFractionDigits` of `20`, and (b) standard notation makes it clearer to the user that the value is an integer.
+    - If the front-end can produce a `number`, then it will use some yet-to-be-determined heuristics to make a best guess as to whether the user would prefer to see the number in standard notation or scientific notation.
+
+### Loss of precision
+
+- To satisfy Goal C, Mathesar delegates responsibility to the back-end for transforming precise values entered by the user into less-precise values stored by Postgres. The front-end does not attempt any rounding or truncation. The back-end responds with the value saved in Postgres and the front-end displays this value. At this point, it is the responsibility of the user to notice that their value has been rounded or truncated. (This behavior is consistent with `psql`, spreadsheets, and AirTable.)
+
+## Future consideration
+
+- At some point we may add display options for numbers, which allow users to customize things like: percentages, minimum fraction digits, maximum fraction digits, parentheses for negative numbers, forcing scientific notation on/off, forcing grouping separators on/off.
```

Comment:
@seancolsen We already have some of these as part of our goals for the alpha. Here's the [figma link](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4260%3A37440&node-id=4308%3A41575&viewport=324%2C48%2C0.35&scaling=min-zoom&starting-point-node-id=4270%3A40031&show-proto-sidebar=1), in which you could find display options within Number.

The display options currently only includes percentages and ability to choose the format from a list of locales. We will not be defaulting to the client's locale all the time. If the client's locale is present in the list we allow, we'd choose it, if not, we'd choose en-us as the default. We would also not have to worry about non-latin numbering systems since we do not intend to support them.

@kgodey Let me know if my assumption is correct on this.

Reply:
We should continue with approach 1. Our plan is for the database layer to be locale-agnostic, and I think that's most in line with our goal of allowing databases created using Mathesar to be used by other clients and tools.

> If the back end is aware of the locale, then the back end can do the formatting and parsing work. 
>
> This is the approach suggested by the Figma specs. The back end knows locale already from a setting persisted at the column level.

Implementing the locale display option for numbers does _not_ mean the backend takes responsibility for formatting and parsing work. Display options are meant to be a cache for the frontend to store preferences, none of the display options (for any data type) affect how the backend works. This is why we have a distinction between "database options" (affects the backend) and "display options" (frontend only).

Implementing the display option would allow the locale used for formatting/parsing to be changed on a per-column basis, but it would still be the frontend's responsibility to use the locale to handle formatting/parsing.

@seancolsen I know when we talked last week, I said we wouldn't be customizing locale per column, I had forgotten about this display option. I don't see how it would change implementation, though. Instead of always defaulting to the browser locale, we would check the locale in the display option and use the browser locale if there's no locale set there.

---

I'm assuming we are in agreement about not supporting non-Latin numbering systems and supporting display options for percentages.

Reply:
@seancolsen 

>  That decision underpins so many other decisions and has the potential to render my entire previous week's worth of work as moot.

I do not see how this is the case. Am I missing something?

As @kgodey mentioned, we would just use a locale stored and sent from the backend and if it's not present we'd default to the browser locale. Additionally, we would also have a list of supported locales, so if the browser locale is not part of that list, we would choose en-us.

The frontend would always be incharge of formatting numbers for display here, and we would interact with the backend only using the canonical format.

Reply:
@kgodey said:

> Display options are meant to be a cache for the frontend to store preferences

I see. Got it. This is sort of a separate hybrid approach, call it `4.`, which builds upon `1.`. Sounds good. 

> I don't see how it would change implementation, though.

Agreed. We can use all the work we've done thus far.

Reply:
@pavish 

> Am I missing something?

I think we're all in agreement on sticking with `1.`, so I don't think it's worth hashing out approaches `2.` or `3.` in great detail. But just to offer some clarity...

The concerns surrounding imports seem to me to weigh in favor of `2.` or `3.` which is why I'm somewhat ambivalent overall and why I've raised this for discussion before continuing my work. If we _did_ decide to go with `2.` or `3.` (parsing and formatting on the back-end), we'd have much less of a need to do parsing and formatting on the front end. If we abandon the goal of formatting numbers _as the user types_, then the front end really could deal with the numerical cells entirely as strings. If that were the case, then I'd probably want to throw out a lot of the code I wrote last week, or maybe port it to Python. That's what I mean by "underpins so many other decisions".